### ISSUE TITLE
nixos-installer error: attribute 'nix-secrets' missing

### ISSUE THREAD

**Issue by EmergentMind** [EmergentMind/nix-config#11]
This is an unintended, but known issue caused by a change to nix-config/vars that introduces values pulled in from the nix-secrets repo via 'inputs.nix-secrets'. `nix-config/nixos-installer/flake.nix` does not use nix-secrets as an input, by design (and necessity as its use is too early for secrets).

There are also a few faulty uses of configVars that leverage values from 'nix-secrets'.

the `just check` recipes have been added, that check both the main flake and the nixos-installer flake are checked to mitigate future over sights of this sort.

Will investigate fix for this.

**Comment by hierocles** [comment]
I'm new to NixOS, so this might not be the best solution. I've managed to pass `just check` with these changes to `vars/default.nix`, along with setting a default SSH port in `nixos-installer/minimal-configuration.nix` and `nixos-installer/iso/default.nix`.

```nix
{
  inputs,
  lib,
  ...
}: let
  # Default value, can be overridden
  defaultIsMinimal = false;

  # Function to get secrets or default values
  getSecrets = isMinimal:
    if !isMinimal && inputs ? nix-secrets
    then inputs.nix-secrets
    else {
      userFullName = null;
      domain = null;
      email = null;
      networking = {
        # Set a custom SSH port here if desired
        # ports = {
        #   tcp = {
        #     ssh = <custom value>;
        #   };
        # };
      };
    };
in rec {
  isMinimal = defaultIsMinimal;

  username = "dylan";
  handle = "hierocles";
  gitHubEmail = "4733259+hierocles@users.noreply.github.com";

  # System-specific settings (FIXME: Likely make options)
  isWork = false; # Used to indicate a host that uses work resources
  scaling = "1"; # Used to indicate what scaling to use. Floating point number

  # Include secrets or default values based on isMinimal
  userFullName = (getSecrets isMinimal).userFullName;
  domain = (getSecrets isMinimal).domain;
  email = (getSecrets isMinimal).email;
  networking = (getSecrets isMinimal).networking;
}
```

In `nixos-installer/minimal-configuration.nix` and `nixos-installer/iso/default.nix`:
```diff
{
  pkgs,
  lib,
  configLib,
  configVars,
  ...
}: let
- sshPort = configVars.networking.ports.tcp.ssh;
+ sshPort = configVars.networking.ports.tcp.ssh or 22;
```

**Comment by EmergentMind** [comment]
Nice, this looks like a good solution. I'll play around with it. Thanks for the suggestion and support!

**Comment by EmergentMind** [comment]
Quick update on this. For unrelated reasons, I'm likely going to be moving away from using configVars in favor of a proper module that provides set of options. A side benefit of doing so is that it will solve this issue. Mentioning for now because the solution proposed above would work nicely but I may not bother implementing it prior to deprecating configVars. Either way, it's on the list for after I wrap up migration of my primary host.

**Comment by DrymarchonShaun** [comment]
> I'm likely going to be moving away from using configVars in favor of a proper module that provides set of options.

I was experimenting with doing this, but I hit a roadblock; It's not possible to use the value of options in imports like we're currently doing with the values of variables in `configVars`, for example [this](https://github.com/DrymarchonShaun/nix-config/commit/7a3c65658701658b8536a15557f12b6060e66b28#diff-a3aa8ffc8ae2b705e38eaf6c11a8fbbd5f8e41366895b081fa97c1e962fea377L12-R14) results in an infinite recursion error because using the option to evaluate something requires everything to have been imported, and for everything to be imported, the value of the option has to be known.

**Comment by EmergentMind** [comment]
Awesome that you've been experimenting with it. The infinite recursion problem is annoying for sure and crops up all over the place. I browsed through your commit and it looks like you had some similar ideas :)
We were able to get past the blocker and I actually finished the last of my refactor yesterday. An example of how we got around it can be seen [here](https://github.com/EmergentMind/nix-config/commit/ec1252cb8fdcfac1837a2b00bb846cf863850792#diff-c1c3fccca11ceffa7e15acd722d6577a371f58d2fdd21f7c11031ac4e2597af1R13)  where some values are declared in let binds first, to be used in imports, and then added to the hostSpec options after the problematic imports are finished. That particular recursion issue only occurred in a couple of files where the hostSpec options are being assigned assembled by common core and the host config.

I'm going to document a few things and update the visual diagram before I squash and merge the branch though as there were some substantial structural modifications. If you're curious there are about 20 commits in my [hostSpec branch](https://github.com/EmergentMind/nix-config/tree/hostSpec) where you can sort of see me stumbling through the various blockers. Lots of other stuff going on in there too that you might be interested in. For example, configLib is now lib.custom and there's dynamic common/core handling for darwin and linux hosts.

**Comment by EmergentMind** [comment]
This issue should be resolved as of 31ab34e, however it hasn't been extensively tested due to #17 
Will track latent nixos-installer issues there.
